src/darwin/Framework/CHIPTests/MTRDeviceTests.m

```diff
@@ -6224,6 +6227,94 @@ - (void)test048_MTRDeviceResubscribeOnSubscriptionPool
+                                                nodeID:@(kDeviceId2)];
+
+    [self waitForExpectations:@[ reachableExpectation ] timeout:60];
+
+    // We relaunch the app multiple times, to try to trigger time
+    // synchronization loss detection. These are the different cases:
+    //
+    //   1) Relaunch with a bad clock, this should just trigger an update,
+    //      because it's the first time that we detect a bad time.
+    //   2) Relaunch with a bad clock again, this should not trigger an update,
+    //      because we avoid doing it too soon after a previous update (see
+    //      MTR_DEVICE_TIME_SYNCHRONIZATION_LOSS_CHECK_CADENCE).
+    //   3) Set the cadence to zero and relaunch with a bad clock again. This
+    //      should trigger an update, because with cadence being zero we won't
+    //      block an update.
+    //   4) Relaunch with a slightly bad clock (and cadence still set to zero).
+    //      This should not trigger an update, because the time is not out of
+    //      sync enough.
+    for (int i = 0; i < 4; ++i) {
+        if (i == 2) {
+            delegate.forceTimeSynchronizationLossDetectionCadenceToZero = YES;
+        }
+
+        XCTestExpectation * subscriptionDroppedExpectation = [self expectationWithDescription:@"Subscription has dropped"];
+        delegate.onNotReachable = ^() {
+            [subscriptionDroppedExpectation fulfill];
+        };
+
+        double utcTime;
+        if (i < 3) {
+            utcTime = 0;
+        } else {
+            utcTime = [[NSDate dateWithTimeIntervalSinceNow:-240] timeIntervalSinceDate:MatterEpoch()];
+        }
+        task = [self relaunchTask:task arguments:@[ @"--use_mock_clock", @(utcTime).stringValue ]];
+
+        [self waitForExpectations:@[ subscriptionDroppedExpectation ] timeout:60];
+
+        XCTestExpectation * resubscriptionReachableExpectation =
+            [self expectationWithDescription:@"Resubscription has become reachable"];
+        delegate.onReachable = ^() {
+            [resubscriptionReachableExpectation fulfill];
+        };
+
+        __weak __auto_type weakDelegate = delegate;
+
+        XCTestExpectation * gotReportsExpectation = [self expectationWithDescription:@"Resubscription got reports"];
+        delegate.onReportEnd = ^() {
+            __strong __auto_type strongDelegate = weakDelegate;
+            strongDelegate.onReportEnd = nil;
+            [gotReportsExpectation fulfill];
+        };
+
+        XCTestExpectation * correctedTime = [self expectationWithDescription:@"onUTCTimeSet called"];
+        delegate.onUTCTimeSet = ^(NSError * _Nullable error) {
+            XCTAssertNil(error);
+            [correctedTime fulfill];
+        };
+        if (i == 1 || i == 3) {
+            correctedTime.inverted = YES;
+        }
+
+        [self waitForExpectations:@[ resubscriptionReachableExpectation, gotReportsExpectation, correctedTime ] timeout:60 enforceOrder:YES];
```

Comment:
When `correctedTime.inverted` is true, does this end up waiting for 60 seconds?  That seems like a long time, if so.  If that's what's going on, then maybe we should give up on the order-checking, wait for resubscriptionReachableExpectation and gotReportsExpectation in order, then wait for correctedTime for a smaller time than 60s (e.g. 5-10s)?  Followup is fine for this.